website/docs/internals/resource-addressing.html.markdown

```diff
@@ -21,12 +21,24 @@ __Module path__:
 A module path addresses a module within the tree of modules. It takes the form:
 
 ```
-module.A.module.B.module.C...
+module.module_name[module index]
 ```
 
-Multiple modules in a path indicate nesting. If a module path is specified
-without a resource spec, the address applies to every resource within the
-module. If the module path is omitted, this addresses the root module.
+ * `module` - Module keyword indicating a child module (non-root). Multiple `module`
+   keywords in a path indicate nesting.
+ * `module_name` - User-defined name of the module.
+ * `[module index]` - (Optional) Index into a module with multiple
+   instances, surrounded by square brace characters (`[` and `]`).
+
+An address without a resource spec, i.e. `module.foo` applies to every resource within
+the module if a single module, or all instances of a module if a module has multiple instances.
+To address all resources of a particular module instance, include the module index in the address,
+such as `module.foo[0]`.
+
+If the module path is omitted, the address applies to the root module.
```

Comment:
Should we also briefly note here that if modules are nested then the `module.name[index]` pattern is repeated once per level?

I don't think we'd need/want to spend a _lot_ of words on it, because deep nesting is something we caution against elsewhere in the docs, but before we had that documented I remember folks trying variants like `module.name1.name2` (without the second `module.` segment) and wondering why it didn't work, so perhaps it would be helpful to show a brief example of the correct syntax for that situation.

Reply:
Yes! I was trying to mirror the brevity this section had before, but I think there could be at least a _little_ more, so that someone would be able to learn about `module.foo[0].module.bar["a"]` as an address, even if we don't want to encourage it.